### ISSUE TITLE
BatchNorm2d implementation returns different results than expected

### ISSUE THREAD

**Issue by dtmoodie** [pytorch/pytorch#20875]
## 🐛 Bug

As per https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d I expect the result of BatchNorm2d to be (x - mean) * scale / sqrt(var + eps) + bias.  However this is not the case.

Given an already loaded BatchNorm2d module as bn0, I have the following:
```
weights = torch.load(model_dir + '29.pt')

model.load_state_dict(weights)
model = model.cuda()
model = model.eval()
base = model.base
# pull out the first batch norm instance in my base feature extractor network.
bn0 = list(list(base.children())[0].children())[1]

bias = bn0.bias[1].item() -> 0.27559953927993774
scale = bn0.weight[1].item() ->  0.4519329071044922
var = bn0.running_var[1].item() -> 2088.339599609375
mean = bn0.running_mean[1].item() -> -4.2851715087890625

n = torch.Tensor(1,32, 2,2)
n[...] = 1.0
y0 = bn0(n.cuda())
y0 = y0[0,1,0,0].item() -> 0.275604248046875
y1 = (1.0 - mean) * scale / math.sqrt(var + bn0.eps) + bias -> 0.32786713069418427
```

## Environment
```
Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.4

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: GeForce GTX 1060 6GB
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

 - PyTorch Version (e.g., 1.0): 1.10
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.5
 - CUDA/cuDNN version: 9.2 / 7.1.3
 - GPU models and configuration: 
 - Any other relevant information:

**Comment by SsnL** [comment]
you are not in eval mode

**Comment by dtmoodie** [comment]
You are incorrect, I call model.eval() in the beginning. I omitted it because that is too obvious, furthermore I've tried leaving that out as well.

**Comment by SsnL** [comment]
I can't reproduce your error.

**Comment by dtmoodie** [comment]
```
bn.state_dict()
Out[7]: 
OrderedDict([('weight',
              tensor([0.46252, 0.45193, 0.35675, 0.21486, 0.73243, 0.42276, 0.27092, 0.66338, 0.44674, 0.47556, 0.69964, 0.38459, 0.58164, 0.61116, 0.46370, 0.59699, 0.81576, 0.43308, 0.29197, 0.58237, 0.32706, 0.04386, 0.31095, 0.50007, 0.43901, 0.34422, 0.51755, 0.72823, 0.31092, 0.54409, 0.33558, 0.50518], device='cuda:0')),
             ('bias',
              tensor([-0.06699,  0.27560,  0.36485, -0.01329,  0.32928,  0.10495,  0.66667, -0.13479, -0.14921, -0.00186, -0.09793, -0.01716,  0.09434,  0.08775,  0.11282,  0.12120, -0.27575,  0.08945, -0.06947,  0.83348,  0.00331, -0.03168, -0.19014,  0.33602,  0.04593, -0.14792, -0.00725,  0.17237,  0.05400,  0.24828,  0.16349, -0.30167], device='cuda:0')),
             ('running_mean',
              tensor([ -4.86915,  -7.15391, -19.52277,  58.62345,  -9.03344,  -6.44255, -28.98025,  36.02863, -10.71378, -41.40824,  57.17245,  53.43351, -31.32692, -42.83203,  21.81406, -38.72557, -87.45533,  51.15632, -63.58952,  -4.54080,  26.08591,  27.14124, -34.84683, -45.23514,  39.87592,  70.51058,  24.64559, -87.98354,   3.79852,   3.31933,   9.08850, 100.26983], device='cuda:0')),
             ('running_var',
              tensor([ 297.33029, 3188.48999, 2126.01514, 1688.14282, 1149.88354,  383.79486, 1535.89624, 1282.59326,  662.05591,  921.21857, 4685.94385, 2387.14062, 1623.79443, 2749.92334,  843.47308, 2624.31030, 4415.36279, 1674.96887, 2039.52893,  169.18286,  830.84766,  398.96970,  712.84973, 1120.70117, 2256.78491, 2187.56250, 1633.21179, 7864.83447,  314.27762,  485.89304,  156.98695, 5104.83301], device='cuda:0')),
             ('num_batches_tracked', tensor(409363, device='cuda:0'))])
```
```
bn = torch.nn.BatchNorm2d(32)
bn.eval()
bn.load_state_dict(state_dict)
bn.track_running_stats = False

def manualBn(bn_op, c):
    bias = bn_op.bias[c].item()
    scale = bn_op.weight[c].item()
    var = bn_op.running_var[c].item()
    mean = bn_op.running_mean[c].item()
    return (1.0 - mean) * scale / (math.sqrt(var + bn0.eps)) + bias

n = torch.Tensor(1,32,2,2)
n[...] = 1.0
y0 = bn(n)[0,:,0,0]
y1 = [manualBn(bn, x) for x in range(32)]

y0
Out[8]: tensor([-0.06699,  0.27560,  0.36485, -0.01329,  0.32927,  0.10495,  0.66667, -0.13478, -0.14922, -0.00186, -0.09793, -0.01716,  0.09435,  0.08775,  0.11281,  0.12120, -0.27576,  0.08946, -0.06947,  0.83348,  0.00331, -0.03168, -0.19015,  0.33603,  0.04593, -0.14793, -0.00725,  0.17236,  0.05400,  0.24828,  0.16348, -0.30167], device='cuda:0', grad_fn=<SelectBackward>)

Out[9]: 
[0.09044372472169493,  0.3408595902259277, 0.5236371016354319, -0.3146224223990977,0.5459921762201441, 0.265558798304041, 0.8739230452927835, -0.7836327506282728, 0.05416714399454639, 0.6626087362818996, -0.6720482709341765,
 -0.42989093156423536, 0.5609540678124897, 0.5985905914664627, -0.2195030143783287,
 0.5841476426518701, 0.8101872265552204, -0.4412895776850876, 0.3480974401349118,
 1.081566720645003, -0.28133327453904355, -0.08908245395532186, 0.2273394338980057,
 1.0266780593210352, -0.3133287800653785, -0.659499264761972, -0.3100660195047065,
 0.903056781221669, 0.004917212697185751, 0.19102890077142384, -0.05315213153664031,
 -1.003564756860683]


```

**Comment by dtmoodie** [comment]
> I can't reproduce your error.

When you say that you cannot reproduce the error, do you get the same output I get from bn?  If so then does your python give you that same correct output when you manually invoke the batch norm equation?

**Comment by SsnL** [comment]
it seems that you have solved this since you closed the issue. but i got the same output as python arithmetic.

**Comment by dtmoodie** [comment]
Yes I've solved the issues and I feel I should appologize for my brash response. The problem was that calling eval() on my model was not propagating to my batch Norm layer. I don't know the reason for that yet but explicitly calling eval on bn0 fixes this.
